### DISCUSSION TITLE
Frequency counting of input signals using the PWMs

### DISCUSSION THREAD

**Discussion by drakejest** [earlephilhower/arduino-pico#1339] [Q&A]
Im reading the pico forums and there seems to be a mention of using the PWMs to be able to get the frequency of an input signals. So far what i have gathered is (no guaranteed if its true):

1. Can count frequency and duty cycle of input signal
2. only works on odd numbered GPIO pins (1,3,5,7,9,11,... etc)

I searched the rp2040 datasheet and came up with something that it might be it 

>[4.5.2.5. Level-sensitive and Edge-sensitive Triggering](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf#page=529)

I think that this must be it as it is mentioned that it only works on PWM B pins (odd numbered GPIO). And the given example even is for measuring duty cycle (but not frequency, maybe can be adjusted to read freqency too). Does this core have an API for this? 

The most important characteristics that i want to know is if this process is non-blocking. If this process of reading the frequency is non-blocking this would be very great for what im currently trying to do. I can free up a lot of clock cycles this way, If not then i guess ill just have to stick with interrupts.

Nonetheless it would be nice if the core has this feature, as i can see a lot of potential use cases even if it is blocking.

**Comment by earlephilhower** [comment]
I don't think the SDK has any APIs for this, but you can do it using direct register access from the `pwm0/1` structs and the datasheet.

It's so asymmetric and limited that I am loathe to add it to the core because there are so many cases when it either can't work due to HW constraints or doesn't come back with the info you think it will.

I'm happy to get a PR, though, if you can come up with a sensible API for it!

**Reply by drakejest** [reply]
Alright then ill give it a go, then come back if ever i make a decent progress :)
